src/build/cargo_plan.rs

```diff
@@ -178,59 +217,18 @@ impl CargoPlan {
             })
             .map(|(key, unit)| (unit.target.src_path().path(), key.clone()))
             .collect();
-        let other_targets: HashMap<UnitKey, &Path> = self
-            .units
-            .iter()
-            .filter(|&(&(_, ref target, _), _)| *target.kind() != TargetKind::CustomBuild)
-            .map(|(key, unit)| {
-                (
-                    key.clone(),
-                    unit.target
-                        .src_path()
-                        .path()
-                        .parent()
-                        .expect("no parent for src_path"),
-                )
-            }).collect();
-
-        for modified in files.iter().map(|x| x.as_ref()) {
-            if let Some(unit) = build_scripts.get(modified) {
-                result.insert(unit.clone());
-            } else {
-                // Not a build script, so we associate a dirty file with a
-                // package by finding longest (most specified) path prefix.
-                let matching_prefix_components = |a: &Path, b: &Path| -> usize {
-                    assert!(a.is_absolute() && b.is_absolute());
-                    a.components().zip(b.components())
-                        .skip(1) // Skip RootDir
-                        .take_while(|&(x, y)| x == y)
-                        .count()
-                };
-                // Since a package can correspond to many units (e.g. compiled
-                // as a regular binary or a test harness for unit tests), we
-                // collect every unit having the longest path prefix.
-                let max_matching_prefix = other_targets
-                    .values()
-                    .map(|src_dir| matching_prefix_components(modified, src_dir))
-                    .max();
-
-                match max_matching_prefix {
-                    Some(0) => error!(
-                        "Modified file {} didn't correspond to any buildable unit!",
-                        modified.display()
-                    ),
-                    Some(max) => {
-                        let dirty_units = other_targets
-                            .iter()
-                            .filter(|(_, dir)| max == matching_prefix_components(modified, dir))
-                            .map(|(unit, _)| unit);
-
-                        result.extend(dirty_units.cloned());
-                    }
-                    None => {} // Possible that only build scripts were modified
-                }
-            }
-        }
+
+        let mut result = HashSet::new();
+
+        let dirty_rustc_units = files
```

Comment:
I should keep the old version for now (it's possible that non-canonical path form mismatch may still not trigger the build correctly?)

Reply:
Reverted for now in https://github.com/rust-lang-nursery/rls/pull/1119/commits/e2c8c2879b57cb7fd4d9865b560ce4f0ca9808f8.